VueJS/src/components/ListItem.vue

```diff
@@ -0,0 +1,58 @@
+<template>
+  <div>
+    <form v-if="editMode">
+      <input type="text" v-model="message" />
+      <button @click="editItem">Save</button>
+      <button @click="editModeEnd">Cancel</button>
+    </form>
+    <p v-if="!editMode">{{item.id}}. {{item.message}}</p>
+    <button @click="editModeStart" v-if="!editMode">Edit</button>
+    <button @click="deleteItem">Delete</button>
+  </div>
+</template>
+
+<script>
+export default {
+  name: "list-item",
+  props: {
+    item: Object
+  },
+  methods: {
+    editModeStart() {
+      this.message = "";
+      this.editMode = true;
+    },
+    editModeEnd() {
+      this.editMode = false;
+    },
+    editItem(e) {
+      e.preventDefault();
+      if (this.message !== "") {
+        this.editMode = false;
+        this.$emit("edit-item", this.$props.item.id, this.message);
```

Comment:
```suggestion
        this.$emit("edit-item", this.todo);
```
See above :point_up:

Reply:
I'm not sure if `this.$emit` is needed when editing an item. As stated in https://vuejs.org/v2/guide/components-props.html#One-Way-Data-Flow: "Note that objects and arrays in JavaScript are passed by reference, so if the prop is an array or object, mutating the object or array itself inside the child component will affect parent state." Therefore, the `editItem` function in _App.vue_ seems unnecessary.

The passed prop to `ListItem` is of type Object within an Array of Objects (`List`). If I run the server with the VueDevTools chrome extension, the `ListItem` indeed updates the parent data in _App.vue_ (and redraws it).

So could I remove that `$emit`?

Reply:
Learned sth. new! Sure, I guess the easiest way is to bring that statement to a test. Remove the `$emit` and see what happens!